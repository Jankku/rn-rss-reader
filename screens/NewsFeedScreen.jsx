import { useContext, useEffect, useState, useRef } from 'react';
import { FlatList, View } from 'react-native';
import Regions from '../data/Regions';
import FeedController from '../data/FeedController';
import NewsItem from '../components/NewsFeed/NewsItem';
import ItemDivider from '../components/NewsFeed/ItemDivider';
import RegionMenuButton from '../components/NewsFeed/RegionMenuButton';
import RegionModal from '../components/NewsFeed/RegionModal';
import { RegionContext } from '../context/RegionContext';
import { useTheme } from '@react-navigation/native';
import ListEmptyComponent from '../components/NewsFeed/ListEmptyComponent';

function NewsFeedScreen({ navigation }) {
  const { region } = useContext(RegionContext);
  const { colors } = useTheme();
  const [newsItems, setNewsItems] = useState([]);
  const [regionModalVisible, setRegionModalVisible] = useState(false);
  const listRef = useRef();

  useEffect(() => {
    (async () => {
      const feed = await FeedController.getFeedById(Regions[region]);
      setNewsItems(feed.rss.channel.item);
    })();
  }, [region]);

  useEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <RegionMenuButton color={colors.headerText} onPress={() => setRegionModalVisible(true)} />
      ),
    });
  }, [colors.headerText, navigation]);

  const _renderItem = ({ item }) => (
    <NewsItem
      title={item?.title}
      description={item?.description}
      imageUrl={item?.enclosure?.url}
      pubDate={item.pubDate}
      onPress={() => navigation.navigate('NewsDetail', { guid: item.guid['#text'] })}
    />
  );

  return (
    <View style={{ flex: 1 }}>
      <FlatList
        ref={listRef}
        data={newsItems}
        keyExtractor={(item) => item.guid['#text']}
        ItemSeparatorComponent={ItemDivider}
        renderItem={_renderItem}
        ListEmptyComponent={<ListEmptyComponent text={"Couldn't find news"} />}
      />

      <RegionModal
        listRef={listRef}
        isVisible={regionModalVisible}
        onClose={() => setRegionModalVisible(false)}
      />
    </View>
  );
}

export default NewsFeedScreen;
