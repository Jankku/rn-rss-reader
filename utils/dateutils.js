import { DateTime } from 'luxon';

/**
 * For articles released today return time and date otherwise.
 * @param {string} pubDate - RFC2822 date
 * @returns {string} Formatted date
 */
export function RFC2822ToTimeOrDate(pubDate) {
  const dateObject = DateTime.fromRFC2822(pubDate);
  const formattedDate =
    dateObject.toRelativeCalendar() === 'today'
      ? dateObject.toLocaleString(DateTime.TIME_24_SIMPLE)
      : dateObject.toLocaleString(DateTime.DATE_SHORT);
  return formattedDate;
}

/**
 * @param {string} pubDate - RFC2822 date
 * @returns {string} Formatted date in DateTime format
 */
export function RFC2822ToDateTime(pubDate) {
  const dateObject = DateTime.fromRFC2822(pubDate);
  const formattedDate = dateObject.toLocaleString(DateTime.DATETIME_SHORT);
  return formattedDate;
}
