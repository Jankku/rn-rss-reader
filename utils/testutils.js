import { render } from '@testing-library/react-native';
import { ThemeContextProvider } from '../context/ThemeContext';
import { LocationContextProvider } from '../context/LocationContext';
import { RegionContextProvider } from '../context/RegionContext';

const AllProviders = ({ children }) => {
  return (
    <LocationContextProvider>
      <RegionContextProvider>
        <ThemeContextProvider>{children}</ThemeContextProvider>
      </RegionContextProvider>
    </LocationContextProvider>
  );
};

const customRender = (ui, options) => render(ui, { wrapper: AllProviders, ...options });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };
