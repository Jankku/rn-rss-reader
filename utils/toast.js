import Toast from 'react-native-root-toast';

const showToast = (message) =>
  Toast.show(message, {
    position: 100,
    opacity: 0.9,
    shadow: false,
  });

export default showToast;
