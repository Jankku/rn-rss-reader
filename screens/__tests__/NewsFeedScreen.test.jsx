import NewsFeedScreen from '../NewsFeedScreen';
import { render } from '../../utils/testutils';

const data = `<rss xmlns:atom="http://www.w3.org/2005/Atom" xmlns:content="http://purl.org/rss/1.0/modules/content/" version="2.0">
<channel>
<language>fi</language>
<title>News</title>
<description>News</description>
<link>https://example.com/</link>
<atom:link href="https://example.com" rel="self" type="application/rss+xml"/>
<category>Tuoreimmat uutiset</category>
<item>
<title>This is a title.</title>
<link>https://example.com</link>
<description>This is a description.</description>
<content:encoded>
Test
</content:encoded>
<pubDate>Fri, 08 Apr 2022 09:19:43 +0300</pubDate>
<category>Category1</category>
<category>Category2</category>
<category>Category3</category>
<guid isPermaLink="false">https://example.com</guid>
</item>
</channel>
</rss>`;

const rssFeed = {
  rss: {
    channel: {
      category: 'Tuoreimmat uutiset',
      description: 'News',
      item: {
        category: ['Category1', 'Category2', 'Category3'],
        description: 'This is a description.',
        encoded: 'This is article content.',
        guid: {
          '#text': 'https://example.com',
          isPermaLink: 'false',
        },
        link: 'https://example.com',
        pubDate: 'Fri, 08 Apr 2022 09:19:43 +0300',
        title: 'This is a title.',
      },
      language: 'fi',
      link: [
        'https://example.com/',
        {
          href: 'https://example.com',
          rel: 'self',
          type: 'application/rss+xml',
        },
      ],
      title: 'News',
    },
    version: '2.0',
  },
};

jest.mock('../../data/FeedController', () => {
  return {
    getFeedById: jest.fn().mockImplementation(() => rssFeed),
  };
});

describe('Test NewsFeedScreen', () => {
  test('Should match snapshot', () => {
    const { toJSON } = render(
      <NewsFeedScreen navigation={{ navigate: () => {}, setOptions: () => {} }} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
