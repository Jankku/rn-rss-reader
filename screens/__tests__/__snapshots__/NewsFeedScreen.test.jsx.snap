// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NewsFeedScreen Should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={
      Object {
        "$$typeof": Symbol(react.memo),
        "compare": null,
        "type": [Function],
      }
    }
    ListEmptyComponent={
      <ListEmptyComponent
        text="Couldn't find news"
      />
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <Text
        style={
          Object {
            "color": "#101010",
            "fontSize": 16,
            "fontWeight": "700",
            "textAlign": "center",
          }
        }
      >
        Couldn't find news
      </Text>
    </View>
  </RCTScrollView>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "position": "absolute",
          "top": 55,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "color": "#101010",
            "elevation": 5,
            "height": 667,
            "shadowOffset": Object {
              "height": 4,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 4,
            "width": 750,
          }
        }
      >
        <Text
          style={
            Object {
              "borderBottomColor": "#dedede",
              "borderBottomWidth": 1,
              "color": "#101010",
              "fontSize": 20,
              "fontWeight": "700",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            }
          }
        >
          Choose region
        </Text>
        <RCTScrollView
          data={
            Array [
              Array [
                "Automatic",
                "automatic",
              ],
              Array [
                "Kotimaa",
                "18-34837",
              ],
              Array [
                "Ulkomaat",
                "18-34953",
              ],
              Array [
                "Lapland",
                "18-139752",
              ],
              Array [
                "North Ostrobothnia",
                "18-148154",
              ],
              Array [
                "Central Ostrobothnia",
                "18-135629",
              ],
              Array [
                "South Ostrobothnia",
                "18-146311",
              ],
              Array [
                "Ostrobothnia",
                "18-148149",
              ],
              Array [
                "Kainuu",
                "18-141399",
              ],
              Array [
                "North Karelia",
                "18-141936",
              ],
              Array [
                "South Karelia",
                "18-141372",
              ],
              Array [
                "North Savo",
                "18-141764",
              ],
              Array [
                "South Savo",
                "18-141852",
              ],
              Array [
                "Central Finland",
                "18-148148",
              ],
              Array [
                "Pirkanmaa",
                "18-146831",
              ],
              Array [
                "Satakunta",
                "18-139772",
              ],
              Array [
                "P채ij채t-H채me",
                "18-141401",
              ],
              Array [
                "Kanta-H채me",
                "18-138727",
              ],
              Array [
                "Kymenlaakso",
                "18-131408",
              ],
              Array [
                "Uusimaa",
                "18-147345",
              ],
              Array [
                "Southwest Finland",
                "18-135507",
              ],
            ]
          }
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  Automatic
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  Kotimaa
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  Ulkomaat
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  Lapland
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  North Ostrobothnia
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  Central Ostrobothnia
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  South Ostrobothnia
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  Ostrobothnia
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  Kainuu
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <Text
                  style={
                    Object {
                      "color": "#101010",
                      "paddingHorizontal": 16,
                      "paddingVertical": 12,
                    }
                  }
                >
                  North Karelia
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "height": 0,
                }
              }
            />
          </View>
        </RCTScrollView>
      </View>
    </View>
  </Modal>
</View>
`;
